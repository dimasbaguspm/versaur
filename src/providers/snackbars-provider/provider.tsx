import { useMemo, type FC } from 'react'
import { SnackbarsContext } from './context'
import type { SnackbarsProviderProps } from './types'

import { Snackbar } from '@/primitive/snackbar'
import {
  snackbarPlacementVariants,
  snackbarTransitionVariants,
} from './helpers'
import { useSnackbarsQueue } from './use-snackbars-queue'

/**
 * SnackbarsProvider manages a queue of snackbars and provides showSnack
 * Placement: full width bottom for mobile, bottom left for desktop/tablet
 */
export const SnackbarsProvider: FC<SnackbarsProviderProps> = ({ children }) => {
  const { queue, showSnack, removeSnack } = useSnackbarsQueue()
  const contextValue = useMemo(() => ({ showSnack }), [showSnack])

  return (
    <SnackbarsContext.Provider value={contextValue}>
      {children}
      <div
        className={snackbarPlacementVariants()}
        aria-live='polite'
        aria-atomic='true'
      >
        {queue.map((snack, idx) => (
          <div
            key={snack.id}
            className={snackbarTransitionVariants({
              placement: window.innerWidth >= 640 ? 'desktop' : 'mobile',
            })}
            style={{ transitionDelay: `${idx * 60}ms` }}
          >
            <Snackbar
              color={snack.color}
              action={snack.action}
              onClose={() => {
                removeSnack(snack.id)
              }}
            >
              {snack.message}
            </Snackbar>
          </div>
        ))}
      </div>
    </SnackbarsContext.Provider>
  )
}
