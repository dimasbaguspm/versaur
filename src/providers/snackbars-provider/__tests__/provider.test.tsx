import { render, screen, act } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../provider.stories'
import { vi } from 'vitest'

const { Basic } = composeStories(stories)

describe('SnackbarsProvider', () => {
  it('renders the provider and children', () => {
    const { asFragment } = render(<Basic />)
    expect(asFragment()).toMatchSnapshot()
    expect(
      screen.getByRole('button', { name: /show success/i })
    ).toBeInTheDocument()
  })

  it('stacks multiple snackbars', () => {
    render(<Basic />)
    act(() => {
      screen.getByRole('button', { name: /show stack/i }).click()
    })
    expect(screen.getByText('Stacked 1')).toBeInTheDocument()
    expect(screen.getByText('Stacked 2')).toBeInTheDocument()
    expect(screen.getByText('Stacked 3')).toBeInTheDocument()
  })

  it('respects custom duration', () => {
    vi.useFakeTimers()
    render(<Basic />)
    act(() => {
      screen.getByRole('button', { name: /show custom duration/i }).click()
    })
    expect(screen.getByText('Custom Duration')).toBeInTheDocument()
    act(() => {
      vi.advanceTimersByTime(4000)
    })

    expect(screen.getByText('Custom Duration')).toBeInTheDocument()
    act(() => {
      vi.advanceTimersByTime(150000)
    })

    expect(screen.queryByText('Custom Duration')).not.toBeInTheDocument()
    vi.useRealTimers()
  })
})
