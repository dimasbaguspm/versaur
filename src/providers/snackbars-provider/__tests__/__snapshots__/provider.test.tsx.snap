// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SnackbarsProvider > renders the provider and children 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-4 p-8"
  >
    <button
      class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none cursor-pointer bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-light focus-visible:ring-offset-white shadow-sm hover:shadow-md h-9 px-4 text-sm min-w-[2.5rem]"
      type="button"
    >
      Show Success
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none cursor-pointer bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-light focus-visible:ring-offset-white shadow-sm hover:shadow-md h-9 px-4 text-sm min-w-[2.5rem]"
      type="button"
    >
      Show Info
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none cursor-pointer bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-light focus-visible:ring-offset-white shadow-sm hover:shadow-md h-9 px-4 text-sm min-w-[2.5rem]"
      type="button"
    >
      Show Warning
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none cursor-pointer bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-light focus-visible:ring-offset-white shadow-sm hover:shadow-md h-9 px-4 text-sm min-w-[2.5rem]"
      type="button"
    >
      Show Danger
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none cursor-pointer bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-light focus-visible:ring-offset-white shadow-sm hover:shadow-md h-9 px-4 text-sm min-w-[2.5rem]"
      type="button"
    >
      Show Stack
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none cursor-pointer bg-primary text-white hover:bg-primary/90 focus-visible:ring-primary-light focus-visible:ring-offset-white shadow-sm hover:shadow-md h-9 px-4 text-sm min-w-[2.5rem]"
      type="button"
    >
      Show Custom Duration
    </button>
  </div>
  <div
    aria-atomic="true"
    aria-live="polite"
    class="fixed z-100 flex flex-col gap-2 px-0 pb-2 w-[80vw] left-0 top-4 md:w-auto md:left-4 md:top-auto md:bottom-4 md:max-w-sm"
  />
</DocumentFragment>
`;
